base_layer/wallet/src/ffi.rs

```diff
@@ -120,284 +435,574 @@ pub unsafe extern "C" fn create_pending_transaction_outputs(
+    transaction: *mut TransactionInput,
 ) -> bool
 {
-    // append this output to PendingTransactionOutputs.outputs_to_be_spent
+    if inputs.is_null() {
+        return false;
+    }
+
+    if transaction.is_null() {
+        return false;
+    }
+
+    (*inputs).0.push((*transaction).clone());
+    return true;
 }
 
-/// Append an UnblindedOutput to be recieved to the pending transaction outputs object
 #[no_mangle]
-pub unsafe extern "C" fn add_output_to_received(
-    pending_tx: *mut PendingTransactionOutputs,
-    value: c_ulonglong,
-    spending_key: *const c_char,
-    feature_flags: c_char, // OutputFlags,
-    maturity: c_ulonglong, // u64
+pub unsafe extern "C" fn transactioninputs_create() -> *mut TransactionInputs {
+    let i = TransactionInputs(Vec::new());
+    Box::into_raw(Box::new(i))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn transactioninputs_destroy(inputs: *mut TransactionInputs) {
+    if inputs.is_null() {
+        let i = Box::from_raw(inputs);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn transactionoutputs_create() -> *mut TransactionOutputs {
+    let i = TransactionOutputs(Vec::new());
+    Box::into_raw(Box::new(i))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn transactionoutputs_destroy(inputs: *mut TransactionOutputs) {
+    if inputs.is_null() {
+        let i = Box::from_raw(inputs);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn transactionkernels_create() -> *mut TransactionKernels {
+    let i = TransactionKernels(Vec::new());
+    Box::into_raw(Box::new(i))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn transactionkernels_destroy(kernels: *mut TransactionKernels) {
+    if kernels.is_null() {
+        let i = Box::from_raw(kernels);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn transactioninput_add_transaction_input(
+    inputs: *mut TransactionInputs,
+    transaction: *mut TransactionInput,
 ) -> bool
 {
-    // append this output to PendingTransactionOutputs.outputs_to_be_received
+    if inputs.is_null() {
+        return false;
+    }
+
+    if transaction.is_null() {
+        return false;
+    }
+
+    (*inputs).0.push((*transaction).clone());
+    return true;
 }
 
-/// Add an output to the wallet. `spent` is a boolean that indicates if this output is a spent or unspent output.
+/// Add a transaction output to a transaction struct
 #[no_mangle]
-pub unsafe extern "C" fn add_pending_transaction_outputs(
-    wallet: *mut WalletFfi,
-    pending_tx: *mut PendingTransactionOutputs,
+pub unsafe extern "C" fn transactionoutputs_add_transaction_output(
+    outputs: *mut TransactionOutputs,
+    transaction: *mut TransactionOutput,
 ) -> bool
 {
-    // append this data to the wallet
+    if outputs.is_null() {
+        return false;
+    }
+
+    if transaction.is_null() {
+        return false;
+    }
+
+    (*outputs).0.push((*transaction).clone());
+    return true;
 }
 
-/// Initialize a Transaction struct to be populated
+/// Add a transaction kernel to a transaction struct
```

Comment:
Same here again. This is why I think we should have a Transaction Obejct rather than breaking it into 3 separate components.